File: lab/01/aquecimentoRecursivo.java
/**
 * Gabriel Xavier Borges - 805347 - Aquecimento01 - 13/08
 */

public class aquecimentoRecursivo
{

    /**
     * Funcao para verificar condicao de parada
     * @param str
     * @return boolean
     */
    public static boolean isEnd (String str)
    {
        boolean r = false;
        if(str.length() == 3 && str.charAt(0) == 'F' && str.charAt(1) == 'I' && str.charAt(2) == 'M')
        {
            r = true;
        }
        return r;
    }

    /**
     * Funcao para verificar se caracter e maiusculo
     * @param c
     * @return boolean
     */
    public static boolean isUpper (char c)
    {
        boolean r = false;
        if(c >= 'A' && c <= 'Z')
        {
            r = true;
        }
        return r;
    }

    /**
     * Funcao para inicializar o contador da funcao countUpper (index)
     * @param str
     * @return index
     */
    public static int countUpper (String str)
    {
        return countUpper(str, 0);
    }

    /**
     * Funcao recursiva para contar letras maiusculas na string
     * @param str
     * @param index
     * @return quantidade
     */
    public static int countUpper (String str, int index)
    {   
        int count = 0;
        if(index < str.length())
        {
            if(isUpper(str.charAt(index)))
            {
                count++;
            }
            count += countUpper(str, index + 1);
        }
        return count;
    }

    /**
     * Funcao principal
     * @param args
     */
    public static void main(String[] args)
    {        
        String str;
        str = MyIO.readLine("");
        do
        {
            System.out.println(""+countUpper(str));
            str = MyIO.readLine("");
        }while(!isEnd(str));
    }
    
}

File: lab/01/aquecimentoRecursivo.c
/**
 * Gabriel Xavier Borges - 805347 - 15/08
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <stdbool.h>

/**
 * Funcao para verificar se e maiuscula
 * @param caracter
 * @return bool
 */
bool isUpper (char c)
{
    bool r = false;
    if(c >= 'A' && c <= 'Z')
    {
        r = true;
    }
    return r;
}

/**
 * Funcao recursiva para contar letras maiusculas numa string
 * @param string, index
 * @return int
 */
int countUpper (char* str, int index)
{
    int count = 0;
    if(index < strlen(str))
    {
        if(isUpper(str[index]))
        {
            count++;
        }
        count += countUpper(str, index + 1);
    }
    return count;
}

/**
 * Funcao auxiliar para chamar a funcao countUpper
 * @param string
 * @return string, index
 */
int countUpperAux (char* str)
{
    return countUpper(str, 0);
}

/**
 * Funcao principal
 */
int main (int agrc, char* agrv[])
{
    char* str = (char*)calloc(80, sizeof(char));

    fgets(str,80,stdin);

    do
    {
        printf("%d\n", countUpperAux(str));
        fgets(str,80,stdin);
    } while (strcmp(str, "FIM\n")!= 0);
    
    free(str);

    return 0;
}